### PR TITLE
Jaskeerat789/issue353

### LINKED ISSUE #353: Feature: Move `region` field from nodepool to provider spec
### Motivation:

Most of the providers define the `region` used on the provider level in terraform, only sometimes on the resource level. 
[GCP](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_reference#example-usage---basic-provider-blocks), [OCI](https://docs.oracle.com/en-us/iaas/Content/API/SDKDocs/terraformproviderconfiguration.htm#APIKeyAuth), [Azure](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/resource_group#location) (under RG spec), [AWS](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#region), Hetzner does not even use `region`, only `zone`

However, in our input manifest, the user is defining the region in the node pool spec, so every node pool has to define it, regardless if it is used or not since we usually look at the `region` of the first node pool. 

I propose to move `region` to provider spec and leave `zone` in node pool spec. If the user would like to use a different region, they would add a provider into their spec, with the same credentials, only a different `region`, as they would have to do it in terraform anyway.

Example:
```yaml
providers:
  aws:
    - name: aws-1
      #some creds
      region: "eu-central-1"
    aws:
    - name: aws-2
      #same creds as in aws-1
      region: "eu-central-2"
  azure:
    - name: azure-1
      ...
      region: "Europe West"
  gcp: 
    - name: gcp-1
      ...
      region: "eu-central1"
  oci:
    - name: oci-1
      ...
      region: "eu-frankfurt-1"

nodePools:
   dynamic:
    - name: pool-1
      providerSpec:
        name: aws-1
        zone: eu-central-1c
      ...
    - name: pool-2
      providerSpec:
        name: oci-1
        zone: hsVQ:EU-FRANKFURT-1-AD-1
      ...
``` 

### Description:

The `region` field should move from node pool spec to provider spec.
### Exit criteria:
- [ ] Change the parsing of the manifest in the frontend/scheduler appropriately
- [ ] Change config proto spec to accommodate the change
- [ ] Change template to accommodate the change

### PR DESCRIPTION
**Pull request by jaskeerat789**
# Description
Fixes #353 by using mukltiple provider definition in terrform scripts. Also updated the test-set to create multi-region clusters.

### CODE REVIEW

File: internal/utils/cluster.go [hunk 1]
```diff
@@ -40,6 +40,24 @@ func GetNodePoolByName(nodePoolName string, nodePools []*pb.NodePool) *pb.NodePo
...
+		regionSet[nodepool.Region] = struct{}{}
+	}
+
+	// extract value of set and create a slice
+	regions := make([]string, len(regionSet))
+	i := 0
+	for k := range regionSet {
+		regions[i] = k
+		i++
+	}
```
Comment (Despire) [inline internal/utils/cluster.go:57]: You request the capacity in the slice and then just append to it
```go
regions := make([]string, 0, len(regionSet))
for k := range regionSet {
regions = append(regions, k)
}
```

File: services/terraformer/server/clusterBuilder/clusterBuilder.go [hunk 2]
```diff
@@ -201,6 +208,12 @@ func (c ClusterBuilder) getNodes() []*pb.Node {
 	return oldNodes
 }
 
+// ReplaceAll is a wrapper function to replace all occurrence of a substring with another
+// substring. Returns modified string
+func (nd NodepoolsData) ReplaceAll(s, old, new string) string {
+	return strings.ReplaceAll(s, old, new)
+}
+
```
Comment (Despire) [inline services/terraformer/server/clusterBuilder/clusterBuilder.go:216]: This function should be part of `NodepoolsData` struct, move the function to this [file](https://github.com/Berops/claudie/blob/ef31f9951082f7d8b2f22df8694884306054f583/internal/templateUtils/templates.go#L1) and then add it [here](https://github.com/Berops/claudie/blob/ef31f9951082f7d8b2f22df8694884306054f583/internal/templateUtils/templates.go#L70) to call it from the templates

Comment (Despire) [reply services/terraformer/server/clusterBuilder/clusterBuilder.go:216]: Actually I think you can just add something like this to the FuncMap [here](https://github.com/Berops/claudie/blob/ef31f9951082f7d8b2f22df8694884306054f583/internal/templateUtils/templates.go#L71) and you shouldn't even need a function definition.
```go
"ReplaceAll": strings.ReplaceAll
```

Comment (jaskeerat789) [reply services/terraformer/server/clusterBuilder/clusterBuilder.go:216]: yeah i was not aware of the function map, is sounds like i dont need to define a special wrapper function, i will pass the `strings.ReplaceAll` in the function map

File: services/terraformer/templates/aws-lb.tpl [hunk 3]
```diff
@@ -1,139 +1,141 @@
...
+{{- $clusterHash := .ClusterHash }}
+{{- $index :=  0 }}
 
+{{- range $i, $region := .Regions }}
 provider "aws" {
-  region     = "{{(index .NodePools 0).Region}}"
-  access_key = "{{(index .NodePools 0).Provider.AwsAccessKey}}"
-  secret_key = file("{{(index .NodePools 0).Provider.SpecName}}")
-  alias = "lb-nodepool"
+  access_key = "{{(index $.NodePools 0).Provider.AwsAccessKey}}"
```
Comment (MiroslavRepka) [inline services/terraformer/templates/aws-lb.tpl:7]: >$.NodePools

why do we need `$`?

File: services/terraformer/templates/hetzner.tpl [hunk 4]
```diff
@@ -29,7 +29,7 @@ resource "hcloud_firewall" "defaultfirewall" {
     ]
   }
 
-  {{ if index .Metadata "loadBalancers" | targetPorts | isMissing 6443 }}
+  {{- if index .Metadata "loadBalancers" | targetPorts | isMissing 6443 }}
```
Comment (MiroslavRepka) [inline services/terraformer/templates/hetzner.tpl:32]: > {{ - if index .Metadata ... }

Why do we need `-` before `if`?

**Review by MiroslavRepka — APPROVED** [review]
LGTM, great job @jaskeerat789 !

**Comment by MiroslavRepka** [comment]
I just realized, since we are no longer supplying `resource_group`, we might as well delete it from `config.proto`

File: services/terraformer/templates/hetzner.tpl (hunk 4 above)
Comment (jaskeerat789) [reply services/terraformer/templates/hetzner.tpl:32]: It's just for cleaner tf files, `-` does not result in new line creation.

**Pull request merged**